docs/design/operator.md

```diff
@@ -0,0 +1,387 @@
+  second instance of a driver of a similar type, it should be deployed in a
+  different namespace.
+- Each CSI driver must have a unique name across all namespaces.
+
+```yaml
+---
+kind: CephCSIDriver
+apiVersion: csi.ceph.io/v1alpha1
+metadata:
+  name: "<prefix>.<driver_type>.csi.ceph.com"
+  namespace: <operator-namespace>
+spec:
+  fsGroupPolicy: File
+  encryption:
+    configMapRef:
+      name: encryption-config-map-name
+  plugin:
+    priorityClassName: system-node-critical
+    updateStrategy:
+      type: RollingUpdate
+      rollingUpdate:
+        maxUnavailable: 1
+    labels:
+      app: cephfs-plugin
+    annotations:
+      k8s.v1.cni.cncf.io/networks: macvlan-conf-1
+  provisioner:
+    labels:
+      app: ceph-fs-provisioner
+    annotations:
+      k8s.v1.cni.cncf.io/networks: macvlan-conf-1
+    provisionerReplicas: 2
+    leaderElection:
+      leaseDuration: 100
+      renewDeadline: 100
+      retryPeriod: 10
+    attachRequired: true
+  liveness:
+    metricsPort: 8000
+  deployCSIAddons: false
+  kernelClient: true
+  kernelMountOptions: ms_mode=secure
+status:
+  phase: Failed
+  reason: csi driver with same name already exists in the cluster
+```
+
+### CephCSICephConfig CRD
+
+Stores connection and configuration details for a single Ceph cluster and
+provide the information to be used by multiple CSI drivers.
+
+```yaml
+---
+kind: CephCSICephCluster
+apiVersion: csi.ceph.io/v1alpha1
+metadata:
+  name: ceph-cluster-1
+  namespace: <operator-namespace>
+spec:
+  monitors:
+  - 10.98.44.171:6789
+  - 10.98.44.172:6789
+  - 10.98.44.173:6789
+  ReadAffinity:
+    crushLocationLabels:
+    - kubernetes.io/hostname
+    - topology.kubernetes.io/region
+    - topology.kubernetes.io/zone
+  rbdMirrorDaemonCount: 2
+  cephConfig:
+    global:
+      auth_cluster_required: none
+      auth_service_required: none
+      auth_client_required: none
+      rbd_validate_pool: false
+status: {}
+```
+
+### CephCSIConfig CRD
+
+Contains details about CephFS, RBD, and NFS configuration to be used when
+communicating with Ceph. Include a reference to a CephCSICephCluster holding
+the connection information for the target Ceph cluster.
+
+```yaml
+---
+kind: CephCSIConfig
+apiVersion: csi.ceph.io/v1alpha1
+metadata:
+  name: storage
+  namespace: <operator-namespace>
+spec:
+  cephClusterRef:
+    name: ceph-cluster-1
+  cephFS:
+    subvolumeGroup: csi
+    kernelMountOptions: readdir_max_bytes=1048576,norbytes
+    fuseMountOptions: debug
+  rbd:
+    radosNamespace: rados-test
+status:
+  phase: Succeeded
+  reason: successfully linked to CephClusterRef
+```
+
+### CephCSIConfigMapping
+
+The CephCSIConfigMapping CR contains a mapping between local and remote Ceph
+cluster configurations. The information in this CR helps Ceph-CSI identify
+peered blockpools. This can help with the correct management and consumption of
+volumes in backup and DR scenarios
+
+```yaml
+---
+kind: CephCSIConfigMapping
```

Comment:
Different question related to peer mapping: why is this a separate CRD and not part of CephCSIDriver? Shouldn't each driver know what its peers are? What is the user story for this being separate versus included in CephCSIDriver?

I feel like this CRD seems strange to be a separate one in the multi-tenant case where a cluster admin wants to allow other users (or sub-admins) to spin up their own CSI drivers.

Reply:
This is the Peer information related to the configuration which Rook is generating here https://github.com/rook/rook/blob/f5ca09e38582d8658491e620e302c174ea11020f/pkg/operator/ceph/csi/peermap/config.go#L51-L110. This doesn't required when cephcsi driver is installed or in most of the normal cases, this is required only for recovery cases, that why its a different CRD.

Reply:
1. This CRD covers cross-driver concerns. 
2. It is a solution to a problem that is not inherited in CSI or its implementation. The problem only manifests itself when we add a layer of cross-cluster replication/mapping of volumes, which today only manifests as a DR-related concern. 

The above two points, and most specifically the second point, highlight that this is not a core concern or a core feature. With that in mind, it made sense to separate into this concern into its own CRD that can evolve on its own and does not impact the evolution of the core CRD as this operator matures